tests/unit/test_wheel.py

```diff
@@ -806,3 +807,36 @@ def test_missing_PATH_env_treated_as_empty_PATH_env(self):
             retval_empty = wheel.message_about_scripts_not_on_PATH(scripts)
 
         assert retval_missing == retval_empty
+
+class TestWheelCommand(object):
+    @patch('pip._internal.wheel.WheelBuilder')
+    @patch('pip._internal.commands.wheel.RequirementSet')
+    def test_save_wheelnames(self, mocked_rs, mocked_wb):
```

Comment:
This test seems like it's going overboard with the mocking, etc, and as a result the test is a lot more complicated than it needs to be, IMO. Ideally, no calls to `mock` or `patch` should be needed. I would suggest changing `save_wheelnames()` to accept `path_to_wheelnames` and `wheel_filenames` and use the `tmpdir` fixture. Then instead of mocking `open()`, just let `save_wheelnames()` write a normal file, and check the file's contents at the end of the test. As for the `RequirementSet` object, I would see what if you can use a `save_wheelnames()` signature that doesn't require mocking there, either. The function only uses `requirement_set.requirements.values()`, so if necessary you could have the function accept just the values. Or if it's easy to construct a `requirement_set.requirements` object, you could have it accept `requirements`. My intention in suggesting a separate function was to have a function that is easy to test while also testing most of the file-writing logic.

Reply:
Feedback of inexperienced contributor:
I admit that I am not familiar with philosophy of tests. My understanding of unit tests is that should check only tested function and should not rely on function calls of others function. When I saw there are directories dedicated to split the types of test(functional, unit) it misled me that the test should be done this way.

Reply:
> Feedback of inexperienced contributor:

@PatrikKoplan The feedback helps. :)

The approach @cjerdonek describes is usually simpler to test on since all the function's "true" dependencies are surfaced and thus the need for mocking is reduced, while making it clearer at the call site what is actually being used.

Do you think that having some prose on this approach to testing would help new contributors understand this better?

Reply:
It might. Either way thanks for guiding me